core/tracker/inclusion.go

```diff
@@ -0,0 +1,300 @@
+	i.mu.Lock()
+	defer i.mu.Unlock()
+
+	for j, sub := range i.submissions {
+		if sub.Included {
+			continue
+		}
+
+		switch sub.Duty.Type {
+		case core.DutyAttester, core.DutyAggregator:
+			_, ok := block.Attestations[sub.AttRoot]
+			if !ok {
+				continue
+			}
+			i.submissions[j].Included = true
+			i.attIncludedFunc(ctx, sub, block)
+		case core.DutyProposer, core.DutyBuilderProposer:
+			if sub.Duty.Slot != block.Slot {
+				continue
+			}
+			i.submissions[j].Included = true
+			// Nothing to report for block inclusions
+		default:
+			panic("bug: unexpected type") // Sanity check, this should never happen
+		}
+	}
+}
+
+// reportMissed reports duties that were broadcast but never included on chain.
+func reportMissed(ctx context.Context, sub submission) {
+	inclusionMisses.WithLabelValues(sub.Duty.Type.String()).Inc()
+
+	switch sub.Duty.Type {
+	case core.DutyAttester, core.DutyAggregator:
+		msg := "Broadcasted attestation never included in any block"
+		if sub.Duty.Type == core.DutyAggregator {
+			msg = "Broadcasted attestation aggregate never included in any block"
+		}
+
+		log.Warn(ctx, msg, nil,
+			z.Any("pubkey", sub.Pubkey),
+			z.I64("attestation_slot", sub.Duty.Slot),
+			z.Any("broadcast_delay", sub.Delay),
+		)
+	case core.DutyProposer, core.DutyBuilderProposer:
+		msg := "Broadcasted block never included in the chain"
+		if sub.Duty.Type == core.DutyBuilderProposer {
+			msg = "Broadcasted blinded block never included in the chain"
+		}
+
+		log.Warn(ctx, msg, nil,
+			z.Any("pubkey", sub.Pubkey),
+			z.I64("block_slot", sub.Duty.Slot),
+			z.Any("broadcast_delay", sub.Delay),
+		)
+	default:
+		panic("bug: unexpected type") // Sanity check, this should never happen
+	}
+}
+
+func reportAttInclusion(ctx context.Context, sub submission, block block) {
+	blockSlot := block.Slot
+	attSlot := int64(block.Attestations[sub.AttRoot].Data.Slot)
+	inclDelay := block.Slot - attSlot
+
+	msg := "Block included attestation"
+	if sub.Duty.Type == core.DutyAggregator {
+		msg += " aggregate"
+	}
+
+	log.Info(ctx, msg,
+		z.I64("block_slot", blockSlot),
+		z.I64("attestation_slot", attSlot),
+		z.Any("pubkey", sub.Pubkey),
+		z.I64("inclusion_delay", inclDelay),
+		z.Any("broadcast_delay", sub.Delay),
+	)
+
+	inclusionDelay.Set(float64(blockSlot - attSlot))
+}
+
+// NewInclusion returns a new InclusionChecker.
+func NewInclusion(ctx context.Context, eth2Cl eth2wrap.Client) (*InclusionChecker, error) {
+	genesis, err := eth2Cl.GenesisTime(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	slotDuration, err := eth2Cl.SlotDuration(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &InclusionChecker{
+		incl: &inclusion{
+			attIncludedFunc: reportAttInclusion,
+			missedFunc:      reportMissed,
+		},
+		eth2Cl:       eth2Cl,
+		genesis:      genesis,
+		slotDuration: slotDuration,
+	}, nil
+}
+
+// InclusionChecker checks whether duties have been included in blocks.
+type InclusionChecker struct {
```

Comment:
why you needed to wrap inclusion with InclusionChecker? Also some methods on `inclusion` struct are public which can be private as they only called by InclusionChecker methods which are exposed to other packages.

Reply:
renamed it to `inclusionChecker` which wrap a `inclusionCore`. The core is decoupled and optimised for easy testing.